### PR TITLE
use image_transport to subscribe to image messages

### LINKED ISSUE #207: Camera display makes unclear assumptions about camera info topic
Currently the camera display assumes that the name of the camera info topic is constructed in the following way:
`<image_topic>/camera_info`

However, this is different from the original rviz which uses `image_transport::getCameraInfoTopic()` to determine the camera info topic.
This is due to the fact that the image_transport package is not availabie for ROS 2 yet.

We should restore the original behavior in time.

### PR DESCRIPTION
**Pull request by ketatam**
ImageDisplay and CameraDisplay now use image_transport to subscribe to image messages, so that it is possible to subscribe to different types of image messages (compressed, theora..).
CameraDisplay also uses image_transport to get the CameraInfoTopic. 
This needs image_common to be added, see corresponding PR: https://github.com/ros2/ros2/pull/884

fixes https://github.com/ros2/rviz/issues/207

### CODE REVIEW

File: rviz_default_plugins/test/rviz_default_plugins/displays/image/get_transport_from_topic_test.cpp
Comment (audrow) [inline rviz_default_plugins/test/rviz_default_plugins/displays/image/get_transport_from_topic_test.cpp:36]: Perhaps you could add [visual tests](https://github.com/ros2/rviz/tree/ros2/rviz_visual_testing_framework) to make sure that different image types are displayed as expected. For example, see [here](https://github.com/ros2/rviz/blob/ros2/rviz_default_plugins/test/rviz_default_plugins/displays/image/image_display_visual_test.cpp).

**Review by audrow — COMMENTED** [review]
Your changes look good to me. However, this PR depends on [`image_common`](https://github.com/ros-perception/image_common/tree/ros2), which is currently not passing CI ([latest attempt](https://travis-ci.org/github/ros-perception/image_common/builds/601511357)) on the `ros2` branch (the branch mentioned in your other PR: https://github.com/ros2/ros2/pull/884). I would feel better about this new dependency if CI was passing on `image_common`'s `ros2` branch.

**Comment by Martin-Idel** [comment]
@audrow Regarding image_common I agree, this should first be in the general ros2.repos file. However, we need new CI - the latest attempt was 10 months ago. In https://github.com/ros2/ros2/pull/884 there was a different CI attempt, but after that, I put in some effort to fix the CI failures, so Jenkins CI was green at one point. Maybe you could give CI another try?

File: rviz_default_plugins/test/rviz_default_plugins/displays/image/get_transport_from_topic_test.cpp
Comment (Martin-Idel) [reply rviz_default_plugins/test/rviz_default_plugins/displays/image/get_transport_from_topic_test.cpp:36]: We discussed this: The problem is that we'd need to publish e.g. compressed images which means we'd need yet another dependency (either opencv or the compressed image publisher) aside from image_common. That's when we decided against adding another visual test - note that there is one for general uncompressed images.

**Comment by audrow** [comment]
Building for CI failed ([Build Status](http://ci.ros2.org/job/ci_linux/11852/)) so I've tried rebasing and rerunning CI:  [Build Status](http://ci.ros2.org/job/ci_linux/11860/).

**Comment by audrow** [comment]
I added the `image_transport` dependency to the `rviz_default_plugins`'s `package.xml`, so that CI will build. 

Here's the new run: 
[Build Status](http://ci.ros2.org/job/ci_linux/11925/)

**Comment by audrow** [comment]
Here's the full CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11945/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6990/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9738/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11890/)

File: rviz_default_plugins/include/rviz_default_plugins/displays/image/image_transport_display.hpp
Comment (Martin-Idel) [inline rviz_default_plugins/include/rviz_default_plugins/displays/image/image_transport_display.hpp]:
```suggestion
#include "image_transport/image_transport.hpp"
#include "image_transport/subscriber_filter.hpp"
```

File: rviz_default_plugins/src/rviz_default_plugins/displays/camera/camera_display.cpp
Comment (Martin-Idel) [inline rviz_default_plugins/src/rviz_default_plugins/displays/camera/camera_display.cpp]:
```suggestion
#include "image_transport/camera_common.hpp"
```

File: rviz_default_plugins/src/rviz_default_plugins/displays/image/get_transport_from_topic.cpp
Comment (Martin-Idel) [inline rviz_default_plugins/src/rviz_default_plugins/displays/image/get_transport_from_topic.cpp:46]:
```suggestion
RVIZ_DEFAULT_PLUGINS_PUBLIC
std::string getTransportFromTopic(const std::string & topic)
```

File: rviz_default_plugins/src/rviz_default_plugins/displays/image/get_transport_from_topic.cpp
Comment (Martin-Idel) [inline rviz_default_plugins/src/rviz_default_plugins/displays/image/get_transport_from_topic.cpp:32]:
```suggestion
#include "rviz_default_plugins/displays/image/get_transport_from_topic.hpp"
#include "rviz_default_plugins/visibility_control.hpp"
```

File: rviz_default_plugins/src/rviz_default_plugins/displays/image/get_transport_from_topic.cpp
Comment (Martin-Idel) [inline rviz_default_plugins/src/rviz_default_plugins/displays/image/get_transport_from_topic.cpp:54]:
```suggestion
RVIZ_DEFAULT_PLUGINS_PUBLIC
std::string getBaseTopicFromTopic(const std::string & topic)
```

**Review by Martin-Idel — CHANGES_REQUESTED** [review]
@audrow Thanks for taking a look. We're probably nearly there.

I don't have push rights to this repository, so I can't simply update this branch. I don't have a Windows machine right here to test, but the problem seems to be just symbol visibility (as always), so this should do the trick.

Also, it's necessary to include the new `.hpp` headers to get rid of warnings (that should fix mac CI).

**Comment by audrow** [comment]
Rerunning after applying changes:
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9754/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11908/)

**Comment by Martin-Idel** [comment]
Ouch, my bad. We have to put the macros to the header files, not the `.cpp` files. That should fix it. Sorry about that.

**Comment by audrow** [comment]
No worries, I should have caught it in reviewing your suggestions. Here's the new Windows run:

Windows [Build Status](http://ci.ros2.org/job/ci_windows/11937/)

**Comment by audrow** [comment]
I'll run CI once more to make sure all OSs build:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12003/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7030/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9781/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11939/)

**Review by audrow — APPROVED** [review]
LGTM, after https://github.com/ros2/ros2/pull/884 is merged in.

Maybe @mjcarroll wants to take a look before this is merged in.

**Comment by Martin-Idel** [comment]
@ketatam https://github.com/ros2/ros2/pull/884 has been merged, do you have some time rebasing this PR and seeing whether it builds? Otherwise I can try to do this over the holidays.

**Comment by vmanchal1** [comment]
Any updates on this? When can this be available to use?

**Comment by Martin-Idel** [comment]
@mjcarroll @audrow

I tested this PR and it still seems to work. Since this is not my repository I cannot merge master into this PR (the conflict is trivial as it's only in the README).

So it should be enough to rebase onto master + new CI for this to get merged. Maybe we can make it happen :).

**Comment by audrow** [comment]
Thanks for the ping, @Martin-Idel. Here's CI after doing a rebase and fixing the merge conflict:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14585/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9340/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12249/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14714/)

**Review by mjcarroll — APPROVED** [review]

**Comment by dwffls** [comment]
Could you show me what the steps would be to implement a compressed image to this visualizer?
Really want to write a PR for this but just don't know where to start.

**Pull request merged**
